backend/auth/firebase.go

```diff
@@ -16,55 +16,92 @@ package auth
+		defer cancel()
+
 		idToken := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
-		userEmail, err := auth.Authenticate(r.Context(), idToken)
+		token, err := auth.FirebaseClient.VerifyIDToken(ctx, idToken)
 		if err != nil {
 			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
-			http.Error(w, err.Error(), http.StatusUnauthorized)
+			http.Error(w, "User authentication failed", http.StatusUnauthorized)
 			return
 		}
-		if getDomain(userEmail) != auth.AuthDomain {
-			http.Error(w, "You are not authorized to view this resource", http.StatusForbidden)
+		if !token.Claims["email_verified"].(bool) {
+			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
+			http.Error(w, "User authentication failed, please verify your email address", http.StatusUnauthorized)
 			return
 		}
-		next(w, r)
+
+		userEmail := token.Claims["email"].(string)
+		user := User{
+			email:  userEmail,
+			domain: getDomain(userEmail),
+		}
+
+		ctxWithUser := context.WithValue(r.Context(), "user", user)
+		next(w, r.WithContext(ctxWithUser))
 	}
 }
 
-func (auth *FirebaseAuth) Authenticate(ctx context.Context, idToken string) (userEmail string, err error) {
-	ctx, cancel := context.WithTimeout(ctx, auth.AuthTimeout)
-	defer cancel()
+func getPermissionsFromTeam(team models.Team, action string) []models.Principal {
+	var permissions []models.Principal
+	if action == ActionRead {
+		permissions = team.Permissions.Read.Allow
+	} else if action == ActionWrite {
+		permissions = team.Permissions.Write.Allow
+	}
+	return permissions
+}
 
-	token, err := auth.FirebaseClient.VerifyIDToken(ctx, idToken)
-	errorMessage := ""
-	if err != nil {
-		errorMessage = "User authentication failed "
-		return "", errors.New(errorMessage)
+func getPermissionsFromGeneral(generalPermissions models.GeneralPermissions, action string) []models.Principal {
+	var permissions []models.Principal
+	if action == ActionReadTeamList {
+		permissions = generalPermissions.ReadTeamList.Allow
+	} else if action == ActionAddTeam {
+		permissions = generalPermissions.AddTeam.Allow
 	}
-	if !token.Claims["email_verified"].(bool) {
-		errorMessage = "User authentication failed, please verify your email address"
-		return "", errors.New(errorMessage)
+	return permissions
+}
+
+func (user User) has(permission models.Principal) bool {
+	return (permission.Type == models.PrincipalTypeDomain && strings.ToLower(permission.ID) == strings.ToLower(user.domain)) ||
+		(permission.Type == models.PrincipalTypeEmail && strings.ToLower(permission.ID) == strings.ToLower(user.email))
+}
+
+func (user User) isPermitted(permissions []models.Principal) bool {
+	isPermitted := false
+	for _, permission := range permissions {
+		if user.has(permission) {
+			isPermitted = true
+		}
 	}
+	return isPermitted
+}
+
+func (auth *FirebaseAuth) CanActOnTeam(user User, team models.Team, action string) bool {
+	permissions := getPermissionsFromTeam(team, action)
```

Comment:
I would rename getPermissionsFromTeam() to getAllowedPrincipalsForTeam() and then call user.isPermitted(allowedPrincipals) which returns whether the user is permitted to perform the action based on the passed allowedPrincipals. Same for canActOnTeamList().

You are not doing anything wrong here. It's just that I feel we can make the code easier to understand with these changes. I would still defer to amedworth@ to make that call.

Reply:
Good point! I've changed it to getTeamAllowedUsers now with users essentially being short for user matcher, the new name I have to the principal. I will check with amedworth@ if this name is okay.